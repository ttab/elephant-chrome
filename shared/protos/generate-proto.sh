#!/usr/bin/env bash

npx protoc \
  --ts_out protos/ \
  --ts_opt generate_dependencies \
  --ts_opt ts_nocheck \
  --ts_opt eslint_disable \
  --proto_path protos \
  protos/service.proto
