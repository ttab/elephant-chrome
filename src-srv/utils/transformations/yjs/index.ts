import * as Y from 'yjs'
import { slateNodesToInsertDelta, yTextToSlateElement } from '@slate-yjs/core'
import type { GetDocumentResponse } from '../../../protos/service.js'
import { type TextbitElement } from '@ttab/textbit'
import { revertNewsdoc } from '../newsdoc/index.js'

export function yjsStateAsUpdate (content: TextbitElement[], data: Y.Doc): Uint8Array {
  const insertContentDelta = slateNodesToInsertDelta(content)
  const sharedContentRoot = data.get('content', Y.XmlText) as Y.XmlText
  sharedContentRoot.applyDelta(insertContentDelta)
  data.share.set('content', sharedContentRoot as Y.AbstractType<any>)

  return Y.encodeStateAsUpdate(data)
}

/**
 * Helper function to transform a newsDoc to a Y.Map()
 * @param data GetDocumentResponse
 * @param map Y.Map<unknown>
 * @returns Y.Map<unknown>
 */
export function newsDocToYmap (data: GetDocumentResponse, map: Y.Map<unknown>): Y.Map<unknown> {
  const d = {
    ...data,
    version: data.version.toString()
  }

  for (const key in d) {
    const value = d[key as keyof typeof d]
    if (typeof value === 'object') {
      const m = new Y.Map()
      if (Array.isArray(value)) {
        const a = new Y.Array()
        if (value?.length > 0) {
          a.push([newsDocToYmap(value as unknown as GetDocumentResponse, m)])
        }
        map.set(key, a)
      } else {
        for (const k in value) {
          m.set(k, value[k as keyof typeof value])
        }
        map.set(key, m)
      }
    }
    map.set(key, value)
  }
  return map
}

/**
* Helper function to transform a Document(Y.Doc) to a newsDoc
* @param document Document
* @returns GetDocumentResponse
*/
export function yDocToNewsDoc (document: Y.Doc): GetDocumentResponse {
  const original = document.get('original', Y.Map) as Y.Map<unknown>
  const json = original.toJSON()

  // revert content
  const sharedRoot = document.get('content', Y.XmlText) as Y.XmlText
  const content = yTextToSlateElement(sharedRoot).children
  json.document.content = revertNewsdoc(content as any)
  json.version = BigInt(json.version)

  return json as GetDocumentResponse
}
