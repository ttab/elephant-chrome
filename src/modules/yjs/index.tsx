export { WebSocketProvider } from './contexts/WebSocketProvider'
