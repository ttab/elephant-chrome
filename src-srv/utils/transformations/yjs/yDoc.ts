import * as Y from 'yjs'
import { type Block, type GetDocumentResponse } from '@/protos/service.js'
import { toYMap } from '../lib/toYMap.js'
import { group, ungroup } from '../lib/group.js'
import { newsDocToSlate, slateToNewsDoc } from '../newsdoc/index.js'
import { slateNodesToInsertDelta, yTextToSlateElement } from '@slate-yjs/core'
import { type TBElement } from '@ttab/textbit'
import { type Document } from '@hocuspocus/server'
import createHash from '../../../../shared/createHash.js'

function yContentToNewsDoc(yContent: Y.XmlText): Block[] | undefined {
  const slateElement = yTextToSlateElement(yContent).children
  return slateToNewsDoc(slateElement as TBElement[])
}

function assertDescriptions(yMeta: Y.Map<unknown>, documentType: string): void {
  // Only perform on planning items and assignments
  if (!['core/planning-item', 'core/assignment'].includes(documentType)) {
    return
  }

  const yDesc = yMeta.get('core/description') as Y.Array<Y.Map<unknown>>

  const payload: Block = {
    id: '',
    uuid: '',
    uri: '',
    url: '',
    type: 'core/description',
    title: '',
    data: { text: '' },
    rel: '',
    role: '',
    name: '',
    value: '',
    contentType: '',
    links: [],
    content: [],
    meta: []
  }

  // No descriptions
  if (!yDesc) {
    const yArr = new Y.Array()
    yArr.push([toYMap({ ...payload, role: 'public' }, new Y.Map())])
    yArr.push([toYMap({ ...payload, role: 'internal' }, new Y.Map())])

    yMeta.set('core/description', yArr)
    return
  }

  // Already has two descriptions
  if (yDesc.length === 2) {
    return
  }

  // Find missing description
  const descriptionTypes = yDesc.map((yMap) => yMap.get('role') as string)
  const missingType = ['public', 'internal'].filter((role) => !descriptionTypes.includes(role))

  missingType.forEach((role) => {
    yDesc.push([toYMap({ ...payload, role }, new Y.Map())])
  })
}

export function newsDocToYDoc(yDoc: Document | Y.Doc, newsDoc: GetDocumentResponse): void {
  try {
    const yMap = yDoc.getMap('ele')
    const { document, version } = newsDoc

    if (document) {
      const { meta, links, content, ...rest } = document

      yMap.set('meta', toYMap(group(document.meta, 'type'), new Y.Map()))
      yMap.set('links', toYMap(group(document.links, 'type'), new Y.Map()))
      yMap.set('root', toYMap(rest, new Y.Map()))

      assertDescriptions(yMap.get('meta') as Y.Map<unknown>, document.type)

      // Assert assignment descriptions
      const yMeta = yMap.get('meta') as Y.Map<unknown>
      const yAssignments = yMeta.get('core/assignment') as Y.Array<unknown>
      if (yAssignments?.length) {
        for (const yAssignment of yAssignments) {
          const assMeta = (yAssignment as Y.Map<unknown>).get('meta') as Y.Map<unknown>
          assertDescriptions(assMeta, 'core/assignment')
        }
      }

      // Share editable content for Textbit use
      const yContent = new Y.XmlText()
      const slateDocument = newsDocToSlate(document?.content ?? [])
      yContent.applyDelta(
        slateNodesToInsertDelta(slateDocument)
      )
      yMap.set('content', yContent)

      // Set version
      const yVersion = yDoc.getMap('version')
      yVersion?.set('version', version?.toString())

      const originalHash = createHash(JSON.stringify(newsDoc.document))

      const yOriginalHash = yDoc.getMap('hash')
      yOriginalHash?.set('hash', originalHash)

      return
    }

    throw new Error('No document')
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message)
    }
    throw new Error('Unknown error')
  }
}

export function yDocToNewsDoc(yDoc: Y.Doc): GetDocumentResponse {
  const yMap = yDoc.getMap('ele')
  try {
    const meta = ungroup((yMap.get('meta') as Y.Map<unknown>)?.toJSON() || {})
    const links = ungroup((yMap.get('links') as Y.Map<unknown>)?.toJSON() || {})

    const yContent = yMap.get('content') as Y.XmlText
    const content = yContent.toString() ? yContentToNewsDoc(yContent) : []


    const root = yMap.get('root') as Y.Map<unknown>

    const { uuid, type, uri, url, title, language } = root.toJSON()

    meta.forEach(docMeta => {
      if (docMeta.type === 'core/assignment') {
        docMeta.meta = docMeta.meta.filter(assMeta => {
          if (assMeta.type === 'core/description') {
            return assMeta.data.text !== ''
          }
          return true
        })
      }
    })

    return {
      version: BigInt(yDoc.getMap('version').get('version') as string),
      document: {
        uuid,
        type,
        uri,
        url,
        title,
        content: content || [],
        // Remove added empty descriptions
        meta: meta.filter((m) => {
          if (m.type === 'core/description') {
            return m.data.text !== ''
          }
          return true
        }),
        links,
        language
      }
    }
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message)
    }
    throw new Error('Unknown error')
  }
}
