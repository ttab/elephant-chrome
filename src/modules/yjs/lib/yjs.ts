import * as Y from 'yjs'
import type { EleDocumentResponse } from '@/shared/types'
import { slateNodesToInsertDelta } from '@slate-yjs/core'

export type YPath = [string, ...(string | number)[]]

type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Record<string, unknown>
    ? DeepPartial<T[P]> // Recursively apply partial updates
    : T[P]
}

/*
 * Create a typed YJS document from a given object.
 */
export function createTypedYDoc(
  data?: EleDocumentResponse,
  options?: {
    document?: Y.Doc
    rootMap?: string
    metaMap?: string
    isReady?: boolean
  }
): Y.Doc {
  // Base document and base maps
  const ydoc = options?.document ?? new Y.Doc()
  const ymap = ydoc.getMap(options?.rootMap ?? 'document')
  const yMeta = ydoc.getMap(options?.metaMap ?? '__meta')

  // Setup __meta (system only information about the document)
  if (options && Object.hasOwnProperty.call(options, 'isReady')) {
    yMeta.set('isReady', options.isReady)
  }

  if (!data?.document) {
    return ydoc
  }

  function populateYMap(obj: EleDocumentResponse, ymap: Y.Map<unknown>) {
    for (const [key, value] of Object.entries(obj)) {
      if (Array.isArray(value)) {
        const yarray = new Y.Array<unknown>()
        value.forEach((item) => {
          if (typeof item === 'object' && item !== null) {
            const childMap = new Y.Map<unknown>()
            populateYMap(item as EleDocumentResponse, childMap)
            yarray.push([childMap])
          } else {
            yarray.push([item])
          }
        })
        ymap.set(key, yarray)
      } else if (typeof value === 'object' && value !== null) {
        const childMap = new Y.Map<unknown>()
        populateYMap(value as EleDocumentResponse, childMap)
        ymap.set(key, childMap)
      } else {
        ymap.set(key, value)
      }
    }
  }

  populateYMap(data, ymap)

  // EleDocument specific handling
  const yContent = new Y.XmlText()
  yContent.applyDelta(
    slateNodesToInsertDelta(data.document.content)
  )

  ymap.set('content', yContent)

  // Set version
  const yVersion = ydoc.getMap('version')
  yVersion?.set('version', data.version)

  return ydoc
}

/*
 * Recursively extracts data from a YJS structure
 */
export function extractYData<T>(y: Y.Map<unknown> | Y.Array<unknown> | Y.XmlText): T {
  if (y instanceof Y.XmlText) {
    return y.toString() as T
  }

  if (y instanceof Y.Map) {
    const result: Record<string, unknown> = {}

    for (const [key, value] of y.entries()) {
      if (value instanceof Y.Map || value instanceof Y.Array) {
        result[key] = extractYData(value)
      } else {
        result[key] = value
      }
    }

    return result as T
  }

  if (y instanceof Y.Array) {
    return y.toArray().map((item) =>
      item instanceof Y.Map || item instanceof Y.Array
        ? extractYData(item)
        : item
    ) as T
  }

  throw new Error('Unsupported Yjs type')
}

/*
 * Recursively updates a YJS document
 *
 * @example
 * updateYMap(document.get('document'), { meta: { priority: Number(value) } })
 */
export function updateYMap<T extends Record<string, unknown>>(ymap: Y.Map<unknown>, newData: DeepPartial<T>) {
  ymap.doc?.transact(() => { // Ensure all changes are atomic
    for (const [key, value] of Object.entries(newData)) {
      if (Array.isArray(value)) {
        const yarray = new Y.Array<unknown>()
        value.forEach((item) => {
          if (typeof item === 'object' && item !== null) {
            const childMap = new Y.Map<unknown>()
            updateYMap(childMap, item as Record<string, unknown>)
            yarray.push([childMap])
          } else {
            yarray.push([item])
          }
        })
        ymap.set(key, yarray)
      } else if (typeof value === 'object' && value !== null) {
        let existingYMap = ymap.get(key) as Y.Map<unknown> | undefined
        if (!(existingYMap instanceof Y.Map)) {
          existingYMap = new Y.Map<unknown>()
          ymap.set(key, existingYMap)
        }
        updateYMap(existingYMap, value as Record<string, unknown>)
      } else {
        ymap.set(key, value)
      }
    }
  })
}

/**
 * Converts a string path to the same array path format used by yjs maintaing
 * the difference between strings for map properties and numbers for array positions.
 *
 * Example:
 * 'meta.myArray[1].value' -> ['meta', 'myArray', 1, 'value]
 */
export function stringToYPath(input: string): YPath {
  const result: Partial<YPath> = []
  const regex = /([^[\].]+)|\[(\d+)\]/g
  let match: RegExpExecArray | null

  while ((match = regex.exec(input)) !== null) {
    if (match[1]) {
      // Matched a word (property in a Y.Map)
      result.push(match[1])
    } else if (match[2]) {
      // Matched a number in brackets (array index)
      result.push(parseInt(match[2]))
    }
  }

  return result as YPath
}

type YjsContainer = Y.Map<unknown> | Y.Array<unknown> | Y.Text | Y.XmlText | Y.XmlFragment | Y.XmlElement

/**
 * Get the path of a Yjs value within its parent structure.
 *
 * @param {Y.AbstractType<unknown>} v - The Yjs value to get the path for.
 * @param {(string | number)[]} path - The current path being constructed.
 * @returns {(string | number)[]} The path of the Yjs value as an array of strings and numbers.
 */
export function getYjsPath(value: YjsContainer | undefined, asString: true): string
export function getYjsPath(value: YjsContainer | undefined, asString?: false): (string | number)[]
export function getYjsPath(value: YjsContainer | undefined, asString: boolean): string | (string | number)[]
export function getYjsPath(
  v?: YjsContainer,
  asString: boolean = false
): (string | number)[] | string {
  const path: (string | number)[] = []

  let current = v
  while (current?._item) {
    const { parent, parentSub } = current._item

    if (parent instanceof Y.Map) {
      path.unshift(parentSub as string)
      current = parent as YjsContainer
    } else if (parent instanceof Y.Array) {
      const index = parent.toArray().indexOf(current)
      path.unshift(index)
      current = parent as YjsContainer
    } else {
      break
    }
  }

  return asString ? path.join('.') : path
}
