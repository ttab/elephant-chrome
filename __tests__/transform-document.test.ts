import { newsDocToYDoc, yDocToNewsDoc } from '../src-srv/utils/transformations/yjs/yDoc'
import * as Y from 'yjs'

import { planning } from './data/planning-newsdoc'
import { article } from './data/article-newsdoc'
import { Block, type GetDocumentResponse } from '@/protos/service'

/*
  * Array order is not guaranteed.
  * Sorts the JSON object recursively so that we can compare the objects
*/
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function sortDocument(json: any): any {
  if (Array.isArray(json)) {
    return json.map(sortDocument)
      .sort((a: unknown, b: unknown) => {
        if (typeof a === 'object' && typeof b === 'object') {
          return JSON.stringify(a).localeCompare(JSON.stringify(b))
        }
        return JSON.stringify(a).localeCompare(JSON.stringify(b))
      })
  } else if (typeof json === 'object' && json !== null) {
    const sortedObject: Record<string, unknown> = {}
    Object.keys(json as Record<string, unknown>).sort().forEach(key => {
      sortedObject[key] = sortDocument((json)[key])
    })
    return sortedObject
  }
  return json
}

describe('Transform full planning newsdoc document to internal YDoc representation', () => {
  const yDoc = new Y.Doc()
  newsDocToYDoc(yDoc, planning)

  it('handles transformation of planning document', () => {
    const planningJson = yDoc.getMap('ele').toJSON()
    expect(planningJson).toMatchSnapshot()
  })


  it('handles reverting the planning document', async () => {
    const { document, version } = await yDocToNewsDoc(yDoc)

    if (!document || !planning.document) {
      throw new Error('no document')
    }

    const augmentedPlanning: GetDocumentResponse = {
      ...planning,
      document: {
        ...planning.document,
        meta: [
          ...(planning.document?.meta || []),
          Block.create({
            type: 'tt/slugline'
          })
        ]
      }
    }

    expect(version).toBe(planning.version)
    expect(sortDocument(document)).toEqual(sortDocument(augmentedPlanning.document))
  })
})

describe('Transform full article newsdoc document to internal YDoc representation', () => {
  const yDoc = new Y.Doc()
  newsDocToYDoc(yDoc, article)

  it('handles article document', () => {
    const articleJson = yDoc.getMap('ele').toJSON()
    expect(articleJson).toMatchSnapshot()
  })


  it('handles reverting the article document', async () => {
    const { document, version } = await yDocToNewsDoc(yDoc)

    if (!document || !article.document) {
      throw new Error('no document')
    }

    expect(version).toBe(article.version)
    expect(sortDocument(document)).toEqual(sortDocument(article.document))
  })
})

describe('Description and slugline handling - planning', () => {
  describe('slugline', () => {
    const yDoc = new Y.Doc()
    newsDocToYDoc(yDoc, planning)

    it('adds slugline to planning and assignment', async () => {
      const sluglineBefore = planning.document?.meta.find((meta) => meta.type === 'tt/slugline')
      expect(sluglineBefore).toBeUndefined()

      const ele = yDoc.getMap('ele')
      const meta = ele.get('meta') as Y.Map<unknown>

      const createdSlugline = (meta.get('tt/slugline') as Y.Map<unknown>).toJSON() as Block[]

      // A slugline is created on the planning document with newsDocToYDoc
      expect(createdSlugline).toEqual([Block.create({ type: 'tt/slugline' })])
      expect(createdSlugline).toHaveLength(1)
      expect(createdSlugline[0].value).toBe('')

      const assignments = (meta.get('core/assignment') as Y.Map<unknown>).toJSON()
      // One slugline each
      expect(assignments[0].meta['tt/slugline']).toHaveLength(1)
      expect(assignments[1].meta['tt/slugline']).toHaveLength(1)

      // First assignment, existing slugline, existing value
      expect(assignments[0].meta['tt/slugline'][0].value).toBe('lands-tomasson')

      // Second assignment, created, empty string
      expect(assignments[1].meta['tt/slugline'][0].value).toBe('')
    })

    it('removes empty sluglines from assignments when reverting', async () => {
      // Revert to newsDoc
      const { document } = await yDocToNewsDoc(yDoc)

      // Created slugline on planning is kept
      const planningSlugline = document?.meta.filter((meta) => meta.type === 'tt/slugline')
      expect(planningSlugline).toHaveLength(1)
      expect(planningSlugline?.[0].value).toBe('')

      const revertedAssignments = document?.meta.filter((meta) => meta.type === 'core/assignment')

      // First assignment slugline is kept
      const firstRevertedAssignment = revertedAssignments?.[0].meta.filter(meta => meta.type === 'tt/slugline')
      expect(firstRevertedAssignment).toHaveLength(1)
      expect(firstRevertedAssignment?.[0].value).toBe('lands-tomasson')

      // Second assignment slugline is removed
      const secondRevertedAssignment = revertedAssignments?.[1].meta.filter(meta => meta.type === 'tt/slugline')
      expect(secondRevertedAssignment).toHaveLength(0)
    })
  })

  describe('Descriptions when empty', () => {
    const yDoc = new Y.Doc()
    newsDocToYDoc(yDoc, planning)

    it('adds two descriptions (public and internal)', () => {
      const planningDescriptions = planning.document?.meta.filter((meta) => meta.type === 'core/description')
      expect(planningDescriptions?.length).toBe(0)

      const meta = yDoc.getMap('ele').get('meta') as Y.Map<unknown>
      const descriptions = meta.get('core/description') as Y.Array<Y.Map<unknown>>
      expect(descriptions.length).toBe(2)
      expect(descriptions.map((d) => d.get('role'))).toEqual(['public', 'internal'])
    })

    it('removes descriptions when reverting', async () => {
      const { document, version } = await yDocToNewsDoc(yDoc)
      if (!document || !planning.document) {
        throw new Error('no document')
      }

      const augmentedPlanning: GetDocumentResponse = {
        ...planning,
        document: {
          ...planning.document,
          meta: [
            ...(planning.document?.meta || []),
            Block.create({
              type: 'tt/slugline'
            })
          ]
        }
      }


      expect(version).toBe(planning.version)
      expect(document.meta.filter((meta) => meta.type === 'core/description').length).toBe(0)
      expect(sortDocument(document)).toEqual(sortDocument(augmentedPlanning.document))
    })
  })

  describe('Description one exists', () => {
    const yDoc = new Y.Doc()
    if (!planning?.document) {
      throw new Error('no document')
    }

    const augmentedPlanning: GetDocumentResponse = {
      ...planning,
      document: {
        ...planning.document,
        meta: [
          ...(planning.document?.meta || []),
          Block.create({
            type: 'core/description',
            data: { text: 'hojhoj' },
            role: 'internal'
          }),
          Block.create({
            type: 'tt/slugline'
          })
        ]
      }
    }

    newsDocToYDoc(yDoc, augmentedPlanning)

    it('adds one when one of other type exists', () => {
      const planningDescriptions = augmentedPlanning.document?.meta.filter((meta) => meta.type === 'core/description')
      expect(planningDescriptions?.length).toBe(1)

      const meta = yDoc.getMap('ele').get('meta') as Y.Map<unknown>
      const descriptions = meta.get('core/description') as Y.Array<Y.Map<unknown>>
      expect(descriptions.length).toBe(2)
      expect(descriptions.map((d) => d.get('role'))).toEqual(['internal', 'public'])
    })

    it('removes them when reverting', async () => {
      const { document, version } = await yDocToNewsDoc(yDoc)
      expect(version).toBe(planning.version)
      expect(document?.meta.filter((meta) => meta.type === 'core/description').length).toBe(1)
      expect(sortDocument(document)).toEqual(sortDocument(augmentedPlanning.document))
    })
  })
})
