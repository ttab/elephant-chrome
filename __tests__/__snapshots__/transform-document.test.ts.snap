// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transform full article newsdoc document to internal YDoc representation > handles article document 1`] = `
{
  "content": "10|000 liter mjölk läckte vid E6En stor mängd mjölk spilldes över vägbanan i samband med en olycka i Falkenberg. Arkivbild.John Doe/TTTrafikOmkring 10|000 liter mjölk läckte ut på vägbanan när en tankbil välte på en avfart från E6 i Falkenberg natten till onsdag, rapporterar Göteborgs-Posten.– Det är många mjölkpaket, säger Peter Karlborg, inre befäl vid räddningstjänsten, till tidningen.Föraren skadades inte i olyckan. Avfarten stängdes av i norrgående riktning medan olyckan hanterades. Den spillda mjölken krävde ingen åtgärd.– Det är ingen tjäle nu och det finns inga vattendrag i närheten, så mjölken går ner i marken bara, säger Karlborg.",
  "links": {
    "core/assignment": [
      {
        "content": {},
        "contentType": "",
        "data": {},
        "id": "",
        "links": {},
        "meta": {},
        "name": "",
        "rel": "assignment",
        "role": "",
        "title": "",
        "type": "core/assignment",
        "uri": "",
        "url": "",
        "uuid": "21543138-859a-49a8-ad9e-2c10dc2d33b1",
        "value": "",
      },
    ],
    "core/content-source": [
      {
        "content": {},
        "contentType": "",
        "data": {},
        "id": "",
        "links": {},
        "meta": {},
        "name": "",
        "rel": "source",
        "role": "",
        "title": "",
        "type": "core/content-source",
        "uri": "tt://content-source/tt",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "core/section": [
      {
        "content": {},
        "contentType": "",
        "data": {},
        "id": "",
        "links": {},
        "meta": {},
        "name": "",
        "rel": "subject",
        "role": "",
        "title": "Inrikes",
        "type": "core/section",
        "uri": "",
        "url": "",
        "uuid": "956636ed-2687-4bc3-a45a-7e09d98c6eeb",
        "value": "",
      },
    ],
  },
  "meta": {
    "core/newsvalue": [
      {
        "content": {},
        "contentType": "",
        "data": {
          "duration": "86400",
          "score": "2",
        },
        "id": "",
        "links": {},
        "meta": {},
        "name": "",
        "rel": "",
        "role": "",
        "title": "",
        "type": "core/newsvalue",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "tt/slugline": [
      {
        "content": {},
        "contentType": "",
        "data": {},
        "id": "",
        "links": {},
        "meta": {},
        "name": "",
        "rel": "",
        "role": "",
        "title": "",
        "type": "tt/slugline",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "mjölkolycka",
      },
    ],
  },
  "root": {
    "language": "sv",
    "title": "mjölkolycka",
    "type": "core/article",
    "uri": "core://article/5a848227-902b-425b-843e-0dd2cd67894b",
    "url": "",
    "uuid": "5a848227-902b-425b-843e-0dd2cd67894b",
  },
}
`;

exports[`Transform full planning newsdoc document to internal YDoc representation > handles transformation of planning document 1`] = `
{
  "content": "",
  "links": {
    "core/section": [
      {
        "content": {},
        "contentType": "",
        "data": {},
        "id": "",
        "links": {},
        "meta": {},
        "name": "",
        "rel": "section",
        "role": "",
        "title": "Sport",
        "type": "core/section",
        "uri": "",
        "url": "",
        "uuid": "a36ff853-9fb3-5950-8893-64ac699f5481",
        "value": "",
      },
    ],
  },
  "meta": {
    "core/assignment": [
      {
        "content": {},
        "contentType": "",
        "data": {
          "end": "2024-02-09T22:59:59Z",
          "end_date": "2024-02-09",
          "full_day": "true",
          "public": "true",
          "publish": "2024-02-09T10:30:00Z",
          "start": "2024-02-08T23:00:00Z",
          "start_date": "2024-02-09",
        },
        "id": "97464374-3d0c-44b9-9126-f5efadbd32a7",
        "links": {
          "core/article": [
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "",
              "rel": "deliverable",
              "role": "",
              "title": "",
              "type": "core/article",
              "uri": "",
              "url": "",
              "uuid": "f283c9a0-6a2e-4021-a009-087961dd032f",
              "value": "",
            },
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "",
              "rel": "deliverable",
              "role": "",
              "title": "",
              "type": "core/article",
              "uri": "",
              "url": "",
              "uuid": "fa5e6fbf-c03d-4447-8a1f-550176979a48",
              "value": "",
            },
          ],
          "core/author": [
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "John Doe/TT",
              "rel": "assignee",
              "role": "primary",
              "title": "",
              "type": "core/author",
              "uri": "",
              "url": "",
              "uuid": "c37fdf3e-72ff-4e22-8b9f-1af0d60b0cd9",
              "value": "",
            },
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "Jane Doe/TT",
              "rel": "assignee",
              "role": "primary",
              "title": "",
              "type": "core/author",
              "uri": "",
              "url": "",
              "uuid": "f23c55de-5e52-428e-9ed7-0cdc0921f60a",
              "value": "",
            },
          ],
        },
        "meta": {
          "core/assignment-type": [
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "",
              "rel": "",
              "role": "",
              "title": "",
              "type": "core/assignment-type",
              "uri": "",
              "url": "",
              "uuid": "",
              "value": "text",
            },
          ],
          "core/description": [
            {
              "content": [],
              "contentType": "",
              "data": {
                "text": "",
              },
              "id": "",
              "links": [],
              "meta": [],
              "name": "",
              "rel": "",
              "role": "public",
              "title": "",
              "type": "core/description",
              "uri": "",
              "url": "",
              "uuid": "",
              "value": "",
            },
            {
              "content": [],
              "contentType": "",
              "data": {
                "text": "",
              },
              "id": "",
              "links": [],
              "meta": [],
              "name": "",
              "rel": "",
              "role": "internal",
              "title": "",
              "type": "core/description",
              "uri": "",
              "url": "",
              "uuid": "",
              "value": "",
            },
          ],
          "tt/slugline": [
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "",
              "rel": "",
              "role": "",
              "title": "",
              "type": "tt/slugline",
              "uri": "",
              "url": "",
              "uuid": "",
              "value": "lands-tomasson",
            },
          ],
        },
        "name": "",
        "rel": "",
        "role": "",
        "title": "Kortare text utsänd. Uppdateras",
        "type": "core/assignment",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
      {
        "content": {},
        "contentType": "",
        "data": {
          "end": "2024-02-09T22:59:59Z",
          "end_date": "2024-02-09",
          "full_day": "true",
          "public": "true",
          "publish": "2024-02-09T10:30:00Z",
          "start": "2024-02-08T23:00:00Z",
          "start_date": "2024-02-09",
        },
        "id": "97464374-3d0c-44b9-9126-f5efadbd32a6",
        "links": {
          "core/article": [
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "",
              "rel": "deliverable",
              "role": "",
              "title": "",
              "type": "core/article",
              "uri": "",
              "url": "",
              "uuid": "f283c9a0-6a2e-4021-a009-087961dd032f",
              "value": "",
            },
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "",
              "rel": "deliverable",
              "role": "",
              "title": "",
              "type": "core/article",
              "uri": "",
              "url": "",
              "uuid": "fa5e6fbf-c03d-4447-8a1f-550176979a48",
              "value": "",
            },
          ],
          "core/author": [
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "John Doe/TT",
              "rel": "assignee",
              "role": "primary",
              "title": "",
              "type": "core/author",
              "uri": "",
              "url": "",
              "uuid": "c37fdf3e-72ff-4e22-8b9f-1af0d60b0cd9",
              "value": "",
            },
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "Jane Doe/TT",
              "rel": "assignee",
              "role": "primary",
              "title": "",
              "type": "core/author",
              "uri": "",
              "url": "",
              "uuid": "f23c55de-5e52-428e-9ed7-0cdc0921f60a",
              "value": "",
            },
          ],
        },
        "meta": {
          "core/assignment-type": [
            {
              "content": {},
              "contentType": "",
              "data": {},
              "id": "",
              "links": {},
              "meta": {},
              "name": "",
              "rel": "",
              "role": "",
              "title": "",
              "type": "core/assignment-type",
              "uri": "",
              "url": "",
              "uuid": "",
              "value": "text",
            },
          ],
          "core/description": [
            {
              "content": [],
              "contentType": "",
              "data": {
                "text": "",
              },
              "id": "",
              "links": [],
              "meta": [],
              "name": "",
              "rel": "",
              "role": "public",
              "title": "",
              "type": "core/description",
              "uri": "",
              "url": "",
              "uuid": "",
              "value": "",
            },
            {
              "content": [],
              "contentType": "",
              "data": {
                "text": "",
              },
              "id": "",
              "links": [],
              "meta": [],
              "name": "",
              "rel": "",
              "role": "internal",
              "title": "",
              "type": "core/description",
              "uri": "",
              "url": "",
              "uuid": "",
              "value": "",
            },
          ],
          "tt/slugline": [
            {
              "content": [],
              "contentType": "",
              "data": {},
              "id": "",
              "links": [],
              "meta": [],
              "name": "",
              "rel": "",
              "role": "",
              "title": "",
              "type": "tt/slugline",
              "uri": "",
              "url": "",
              "uuid": "",
              "value": "",
            },
          ],
        },
        "name": "",
        "rel": "",
        "role": "",
        "title": "Assignment 2",
        "type": "core/assignment",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "core/description": [
      {
        "content": [],
        "contentType": "",
        "data": {
          "text": "",
        },
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "public",
        "title": "",
        "type": "core/description",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
      {
        "content": [],
        "contentType": "",
        "data": {
          "text": "",
        },
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "internal",
        "title": "",
        "type": "core/description",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "core/newsvalue": [
      {
        "content": {},
        "contentType": "",
        "data": {},
        "id": "",
        "links": {},
        "meta": {},
        "name": "",
        "rel": "",
        "role": "",
        "title": "",
        "type": "core/newsvalue",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "4",
      },
    ],
    "core/planning-item": [
      {
        "content": {},
        "contentType": "",
        "data": {
          "end_date": "2024-02-09",
          "public": "true",
          "start_date": "2024-02-09",
          "tentative": "false",
        },
        "id": "",
        "links": {},
        "meta": {},
        "name": "",
        "rel": "",
        "role": "",
        "title": "",
        "type": "core/planning-item",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
    "tt/slugline": [
      {
        "content": [],
        "contentType": "",
        "data": {},
        "id": "",
        "links": [],
        "meta": [],
        "name": "",
        "rel": "",
        "role": "",
        "title": "",
        "type": "tt/slugline",
        "uri": "",
        "url": "",
        "uuid": "",
        "value": "",
      },
    ],
  },
  "root": {
    "language": "sv-se",
    "title": "Tomasson lämnar Blackburn – för landslaget?",
    "type": "core/planning-item",
    "uri": "core://newscoverage/726045e7-a52b-4737-8fba-c8149f7e2c2d",
    "url": "",
    "uuid": "726045e7-a52b-4737-8fba-c8149f7e2c2d",
  },
}
`;
